sdk/spring/spring-cloud-azure-feature-management/src/main/java/com/azure/spring/cloud/feature/management/filters/TargetingFilter.java

```diff
@@ -111,26 +119,45 @@ public boolean evaluate(FeatureFilterEvaluationContext context) {
 
         Map<String, Object> parameters = context.getParameters();
 
-        if (parameters != null) {
-            Object audienceObject = parameters.get(AUDIENCE);
-            if (audienceObject != null) {
-                parameters = (Map<String, Object>) audienceObject;
-            }
+        Object audienceObject = parameters.get(AUDIENCE);
+        if (audienceObject != null) {
+            parameters = (Map<String, Object>) audienceObject;
+        }
 
-            updateValueFromMapToList(parameters, USERS);
-            updateValueFromMapToList(parameters, GROUPS);
+        updateValueFromMapToList(parameters, USERS);
+        updateValueFromMapToList(parameters, GROUPS);
 
-            settings.setAudience(OBJECT_MAPPER.convertValue(parameters, Audience.class));
+        Map<String, Map<String, Object>> exclusionMap = (Map<String, Map<String, Object>>) parameters
+            .get(EXCLUSION);
+        Map<String, Object> exclusionAsLists = new HashMap<>();
+        if (exclusionMap != null) {
+            exclusionAsLists.put(USERS, exclusionMap.getOrDefault(USERS, new HashMap<String, Object>()).values());
+            exclusionAsLists.put(GROUPS, exclusionMap.getOrDefault(GROUPS, new HashMap<String, Object>()).values());
         }
+        parameters.put(EXCLUSION, exclusionAsLists);
+        settings.setAudience(OBJECT_MAPPER.convertValue(parameters, Audience.class));
 
         validateSettings(settings);
 
         Audience audience = settings.getAudience();
 
-        if (targetingContext.getUserId() != null
-            && audience.getUsers() != null
-            && audience.getUsers().stream()
-                .anyMatch(user -> equals(targetingContext.getUserId(), user))) {
+        // Need to Check denied first
+        if (targetUser(targetingContext.getUserId(), audience.getExclusion().getUsers())) {
+            return false;
+        }
+
+        if (targetingContext.getGroups() != null && audience.getExclusion().getGroups() != null) {
+            for (String group : targetingContext.getGroups()) {
+                Optional<String> groupRollout = audience.getExclusion().getGroups().stream()
+                    .filter(g -> equals(g, group)).findFirst();
+                if (groupRollout.isPresent()) {
+                    return false;
+                }
+            }
+        }
```

Comment:
Isn't this the same logic as targetUser? Can we have a targetGroup method here?

Reply:
Users 100% get a feature. A percentage of a group can get a feature. i.e Ring1 users get it 95% of the time and Ring2 users get it 25% of the time.

Reply:
ok, so that's the findFirst. But can we still extract this into a method?